### PR TITLE
V2: Rename model output file to clarify best vs. last model

### PR DESCRIPTION
**Pull request by hwpang**
## Description
I rename the checkpoint files to be explicit about whether they are the best model (based on validation set) or the last model.

I also save a model file version of the best model by loading the best checkpoint. To me, the whole point of having the model file is so that the inference can use a lightweight file. I don't see when we would want to run inference from the last model but not the best model. I think we can just remove the `last.pt`.

## Example / Current workflow
n/a

## Bugfix / Desired workflow
- Renamed checkpoints to reflect whether they are the last model or the best model
- Save a model file version of the best model
- Change all scripts, tests, and the documentation to point to the `best.pt`.

## Questions
n/a

## Relevant issues
n/a

## Checklist
- [ ] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

File: chemprop/cli/train.py [hunk 1]
```diff
@@ -864,7 +864,7 @@ def train_model(
         trainer.fit(model, train_loader, val_loader)
 
         if test_loader is not None:
-            predss = trainer.predict(model, test_loader)
+            predss = trainer.predict(test_loader)
```
Comment (shihchengli) [inline chemprop/cli/train.py:867]: To clarify, does this line mean that it will be tested on the best model?

File: chemprop/cli/train.py [hunk 2]
```diff
@@ -920,9 +920,15 @@ def train_model(
                 df_preds = pd.DataFrame(list(zip(*namess, *preds.T)), columns=columns)
             df_preds.to_csv(model_output_dir / "test_predictions.csv", index=False)
 
-        p_model = model_output_dir / "model.pt"
+        p_model = model_output_dir / "last.pt"
         save_model(p_model, model)
-        logger.info(f"Model saved to '{p_model}'")
+        logger.info(f"Model from last epoch saved to '{p_model}'")
```
Comment (shihchengli) [inline chemprop/cli/train.py:925]:
```suggestion
        logger.info(f"Model from the last epoch saved to '{p_model}'")
```

File: tests/regenerate_models.sh [hunk 3]
```diff
@@ -16,23 +16,23 @@ rm -rf test_cli_classification_mol
...
 # test_cli_regression_mol+mol
 
 rm -rf test_cli_regression_mol+mol
 
 chemprop train -i $CHEMPROP_PATH/tests/data/regression/mol+mol/mol+mol.csv --accelerator cpu --epochs 1 --num-workers 0 --smiles-columns smiles solvent --save-dir test_cli_regression_mol+mol
 
-cp -L test_cli_regression_mol+mol/model_0/model.pt $CHEMPROP_PATH/tests/data/example_model_v2_regression_mol+mol.pt
+cp -L test_cli_regression_mol+mol/model_0/best.pt $CHEMPROP_PATH/tests/data/example_model_v2_regression_mol+mol.pt
 
 cp -L test_cli_regression_mol+mol/model_0/checkpoints/last.ckpt $CHEMPROP_PATH/tests/data/example_model_v2_regression_mol+mol.ckpt
```
Comment (shihchengli) [inline tests/regenerate_models.sh:37]: Change `last.ckpt` to the `best-{epoch}-{val_loss:.2f}.ckpt`?

File: tests/regenerate_models.sh [hunk 4]
```diff
@@ -42,7 +42,7 @@ rm -rf test_cli_regression_mol
 
 chemprop train -i $CHEMPROP_PATH/tests/data/regression/mol/mol.csv --accelerator cpu --epochs 1 --num-workers 0 --save-dir test_cli_regression_mol
 
-cp -L test_cli_regression_mol/model_0/model.pt $CHEMPROP_PATH/tests/data/example_model_v2_regression_mol.pt
+cp -L test_cli_regression_mol/model_0/best.pt $CHEMPROP_PATH/tests/data/example_model_v2_regression_mol.pt
 
 cp -L test_cli_regression_mol/model_0/checkpoints/last.ckpt $CHEMPROP_PATH/tests/data/example_model_v2_regression_mol.ckpt
```
Comment (shihchengli) [inline tests/regenerate_models.sh:47]: same here

File: tests/regenerate_models.sh [hunk 5]
```diff
@@ -52,22 +52,22 @@ rm -rf test_cli_regression_mol_multitask
...
+cp -L test_cli_regression_rxn+mol/model_0/best.pt $CHEMPROP_PATH/tests/data/example_model_v2_regression_rxn+mol.pt
 
 # test_cli_regression_rxn
 
 rm -rf test_cli_regression_rxn
 
 chemprop train -i $CHEMPROP_PATH/tests/data/regression/rxn/rxn.csv --accelerator cpu --epochs 1 --num-workers 0 --reaction-columns smiles --save-dir test_cli_regression_rxn
 
-cp -L test_cli_regression_rxn/model_0/model.pt $CHEMPROP_PATH/tests/data/example_model_v2_regression_rxn.pt
+cp -L test_cli_regression_rxn/model_0/best.pt $CHEMPROP_PATH/tests/data/example_model_v2_regression_rxn.pt
```
Comment (shihchengli) [inline tests/regenerate_models.sh:71]: same here

**Review by KnathanM — APPROVED** [review]

File: tests/cli/test_cli_regression_mol+mol.py [hunk 6]
```diff
@@ -131,7 +131,7 @@ def test_train_output_structure(monkeypatch, data_path, tmp_path):
         m.setattr("sys.argv", args)
         main()
 
-    assert (tmp_path / "model_0" / "model.pt").exists()
+    assert (tmp_path / "model_0" / "best.pt").exists()
```
Comment (shihchengli) [inline tests/cli/test_cli_regression_mol+mol.py:134]: Additionally, should we add `assert (tmp_path / "model_0" / "last.pt").exists()` in tests? Although, probably no one needs the last file though.

**Comment by KnathanM** [comment]
You just need to change `predss = trainer.predict(test_loader)` to `predss = trainer.predict(dataloaders=test_loader)`, I think

**Review by shihchengli — COMMENTED** [review]
I think we can just remove the last.pt.

File: chemprop/cli/train.py (hunk 1 above)
Comment (KnathanM) [reply chemprop/cli/train.py:867]: I ran some tests and can confirm that if the trainer isn't given a model, it will use the best validation score checkpoint model from the previous `trainer.fit` (assuming that they occur in the same function call as they do here).

File: tests/regenerate_models.sh (hunk 3 above)
Comment (KnathanM) [reply tests/regenerate_models.sh:37]: What is hard about that is the name is dynamically made by lightning. We could use `.../model_0/checkpoint/best*.ckpt`. Or we could use `last.pt` and `last.ckpt` as these are trained for 1 epoch anyways. But if we want to not save `last.pt` (fine by me), then using a glob (`*`) is probably the best route. We have some tests where we check that the .pt and .ckpt give the same prediction, so we need to be consistent.

File: tests/cli/test_cli_regression_mol+mol.py (hunk 6 above)
Comment (hwpang) [reply tests/cli/test_cli_regression_mol+mol.py:134]: I will just remove `last.pt`

**Review by shihchengli — APPROVED** [review]
